Changes to  wxStyledTextCtrl related to XPM images

NPDevelopmentCode opened the pull request:
This fixes 2 issues related to XPM images in wxStyledTextCtrl:

1. According to the [Scintilla specification](https://www.scintilla.org/ScintillaDoc.html#Images), calling `SendMsg` with `SCI_REGISTERIMAGE` should accept both a copy of an XPM file and a set of XPM data.  Currently it only accepts the file form.
2. The `wxStyledTextCtrl::MarkerDefineBitmap` and `wxStyledTextCtrl::RegisterImage` methods currently require the XPM image handler even if no XPM images are being used.

Because of the way the second issue is fixed, markers and autocompletion images can now use full alpha data instead of just transparent and opaque.

Unfortunately since the wxSTC files are partly autogenerated, these fixes ended up changing 10 files.  I split this into 6 separate commits to help track what is being changed at each step.

Commit:
Preserve recent changes to wxSTC documentation

Recent commits 8fbca5cb70 and c68e5d0617 fixed some typos is the
documentation for wxSTC. Unfortunatly these changes will be lost the
next time the wxSTC files are regenerated. This commit modifies
stc.interface.h.in and gen_docs.py to ensure that these changes are
preserved when the wxSTC files are regenerated.

Commit:
Implement BitmapFromRGBAImage without raw bitmap

The function BitmapFromRGBAImage in PlatWX.cpp is currently only
used when wxHAS_RAW_BITMAP is defined. Consequently some functions in
PlatWX.cpp simply do nothing when wxHAS_RAW_BITMAP is not defined. This
provides a second implementation of the function based on wxImage that
is used when wxHAS_RAW_BITMAP is not defined.

vadz commented on src/stc/stc.cpp.in:
```diff
@@ -563,6 +564,41 @@ void wxStyledTextCtrl::AnnotationClearLine(int line) {
+void wxStyledTextCtrl::MarkerDefineBitmap(int markerNumber, const wxBitmap& bmp)
+{
+    // Store the current RGBAImage width and height.
+    const int oldRGBWidth = m_swx->GetRGBAImageWidth();
```
I'm a bit confused by this: why do we need to save and restore this width/height? Is there a "current RGBA image" in STC?

NewPagodi replied:
Scintilla keeps track of the width and height used for registering markers and images with RGBA data.  The specific variable used is the `Point sizeRGBAImage;` member from the editor class.

I thought it was important to save and restore for a scenario like this. Suppose someone has existing code registering 3 markers with 3 sets of 10x10 rgba data. 

```
m_stc->RGBAImageSetWidth(10);
m_stc->RGBAImageSetHeight(10);
m_stc->MarkerDefineRGBAImage(1,rgba1);
m_stc->MarkerDefineRGBAImage(2,rgba2);
m_stc->MarkerDefineRGBAImage(3,rgba3);
```

But suppose they later want to use a 12x12 png for the second marker. They don't want to go though the trouble of extracting the rgba data from the png, so they just load it into wxBitmap `bmp` and change the code to 

```
m_stc->RGBAImageSetWidth(10);
m_stc->RGBAImageSetHeight(10);
m_stc->MarkerDefineRGBAImage(1,rgba1);
m_stc->MarkerDefineBitmap(2,bmp);
m_stc->MarkerDefineRGBAImage(3,rgba3);
```

Since `MarkerDefineBitmap` internally calls `RGBAImageSetWidth/Height(12)`, the call `MarkerDefineRGBAImage(3,rgba3)` will now produce a garbled image (and possibly crash). Since a user might not expect `MarkerDefineBitmap` to change the current width and height used by `MarkerDefineRGBAImage`, I thought it best to save and restore them.

This is one of the reasons I was trying to come up with a way to pass a wxBitmap straight through in the `MarkerDefineRGBAImage`/`ResisterImage` methods. But never found a solutions that wasn't too ugly to use.

vadz replied:
I see, thanks for the explanation. It looks a bit weird to me to rely on width/height applying to more than just the next `DefineRGBAImage()` call, but it's better to be safe, thanks.

vadz commented on src/stc/stc.interface.h.in:
```diff
@@ -353,6 +353,11 @@ public:
 
+    /**
+       Define a marker from a bitmap.
+    */
```
There should probably be `@since 3.1.3` here.

vadz commented:
I don't fully understand everything here, but I'll still merge this unless anybody objects in the next couple of days, thanks for your work on this!

One other question I have is whether it's really worth to have both raw bitmap and `wxImage` versions of the code: if we need the latter anyhow (why, BTW? i.e. which port are you using that it doesn't support raw bitmap pixel access?), I think we could just use it always as performance difference shouldn't be important anyhow for the typically small bitmaps used here.

Thanks again!

vadz commented:
> 
> 
> I wasn't sure if there are any port that did not support raw bitmap access.

All the major ones do support them and all the others almost certainly have worse problems than lack of marker bitmap support with wxSTC too...

> But since I changed `RegisterImage(int, const wxBitmap&)` to call `RegisterRGBAImage`, I wanted to be 100% sure that `RegisterRGBAImage` will work.
> 
> I can change [91d3981](https://github.com/wxWidgets/wxWidgets/commit/91d3981a7bfb4de206fea4a4f89ed07d57b269e8) to completely replace the BitmapFromRGBAImage function if you want. If I do that, should I also rewrite the the `SurfaceImpl::AlphaRectangle` method and completely remove the items using raw bitmap pixel data.

I think it's better to keep things (as) simple (as possible), but I don't really insist on removing these versions. So we could merge this PR as it is, and if you feel like simplifying it later, please make another one with this change (of course, if you prefer to do it right now, please feel free to do it too).

Thanks again!

NewPagodi commented:
I think I have 2 improvements that will avoid the issue with saving and restoring Scintilla's rgba size mentioned above. 

----

First I think we can redefine `MarkerDefineBitmap` to call a utility function in the the `ScintillaWX` class declared like this:

```
void ScintillaWX::MarkerDefineBitmapHelper(int markerNumber,
                                           const unsigned char* data,
                                           int width, int height, int scale)
```
This helper function can use mostly the same code Scintilla uses to handle the `SCI_MARKERDEFINERGBAIMAGE` message except with the values for width and height passed to the function instead of the values it's storing. 

In addition, I think MarkerDefineBitmap should be given an optional 3rd parameter that can set a scale for the marker (100 = regular size, 200 = double size, etc).  The reason for this is that `MarkerDefineRGBAImage` method can use a scale factor defined with the `RGBAImageSetScale` method, and `MarkerDefineBitmap` should probably offer the same functionality.

Incidentally, scaled rgba markers aren't currently being drawn correctly if their scale is not 100, but I'll address that in a subsequent pull request.

----

Second, I think `RegisterImage` can be rewritten to call another utility function

```
void wxStyledTextCtrl::RegisterImage(int type, const wxBitmap& bmp) {
    m_swx->RegisterImageHelper(type, bmp);
}
```
where RegisterImageHelper is a method of ScintillaWX defined as follows:

```
void ScintillaWX::RegisterImageHelper(int type, const wxBitmap& bmp)
{
    static_cast<ListBoxImpl*>(ac.lb)->RegisterImageHelper(type, bmp);
}
```
`ac` is a member of ScintillaWX containing autocompletion data and `lb` is a member of `ac` that is a pointer to an object of the ListBox class defined in Scintilla's platform.h header.  But with wxSTC, we use a derived class ListBoxImpl instead - and that class already has the method needed.

This has the added benefit that we don't need to convert the bitmap to rgba data and then convert it straight back to a bitmap.  The potential drawback here is that in order for this cast to work, I'll have to move the declaration of the `ListBoxImpl` class from PlatWX.cpp to PlatWX.h. But actually I'd like to do that anyway to help with some additional changes for autocompletion popups in a subsequent pull request.

----

A potential drawback with both of these is that they make essential use of the Scintilla objects of the ScintillaWX class, and the code might be unfamiliar to subsequent maintainers.  But other methods in ScintillaWX do as well, so I hope this is OK. 

If there are no objections, I'll update this pull request in a day or so.

Commit:
Change wxSTC mapping for SCI_MARKERDEFINEPIXMAP

Currently the Scintilla message SCI_MARKERDEFINEPIXMAP is mapped to the
wxStyledTextCtrl::MarkerDefineBitmap method. This has two
drawbacks. First this requires the XPM image handler be loaded before
this method can be called. Second, any alpha data except for opaque and
transparent is lost in the conversion to XPM format.

Instead have MarkerDefineBitmap be a manually declared method but
reimplemented it in a way similar to how the SCI_MARKERDEFINERGBAIMAGE
message works. The new implementation preserves alpha data if it exists.

To backfill the message map, the SCI_MARKERDEFINEPIXMAP is now mapped to
a new method MarkerDefinePixmap(int, const char* const*). The new method
accepts XPM data instead of a wxBitmap.

NewPagodi commented:
I've updated this pull request with the changes mentioned above. I think this is a more technically correct way to implement the MarkerDefineBitmap and RegisterImage methods.

Commit:
Change the implementation of ListBoxImpl::RegisterImage

The ListBoxImpl::RegisterImage in PlatWX.cpp is supposed to accept an
XPM and convert it into a useable form. For wxWidgets, the useable form
is obviously a wxBitmap. According to the Scintilla specification, the
function should accept both a copy of an XPM file and a set of XPM data.

Currently RegisterImage uses the the wxWidgets XPM image handler. This
has 2 drawbacks. First it requires that the XPM handler is loaded before
the function can be called. Second, the function only accepts a copy of
an XPM file and does not work with XPM data.

Instead use wxXPMDecoder. This class can be decode both types of input
and can be used to build a wxBitmap.

Commit:
Change wxSTC mapping for SCI_REGISTERIMAGE

Currently the Scintilla message SCI_REGISTERIMAGE is mapped to
wxStyledTextCtrl::RegisterImage(int, const wxBitmap&). This makes
RegisterImage a manually defined method and passes the bitmap directly
to the listbox instead of first converting to an XPM.

To backfill the message map, SCI_REGISTERIMAGE is now mapped to a new
method overload RegisterImage(int, const char* const*). The new method
accepts XPM data instead of a wxBitmap.

Commit:
Regenerate wxSTC files after recent changes

NewPagodi commented:
I had to update the third commit of this pull request to change the implementation of ListBoxImpl::RegisterImage because of a issue on macos. For some reason, when XPM images are typed directly into the file, they don't work as intended.  The new implementation works if the XPM is typed directly into the source file or included with the #include statement.  I've made a post with a slightly longer discussion and some comparison images of the dev list if that would be helpful.

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -2541,20 +2521,31 @@ void ListBoxImpl::RegisterImageHelper(int type, wxBitmap& bmp)
+    wxXPMDecoder dec;
+    wxImage img;
+
+    // This check is borrowed from src/stc/scintilla/src/XPM.cpp.
```
As mentioned on the mailing list, could we please replace this with `strncmp(xpm_data, "/* XPM */", 9) == 0`? Or am I missing some reason for which this simpler test wouldn't work?

NewPagodi replied:
I believe the idea behind that check is that the function can be either passed a set of XPM data, a C array of C strings, or a NULL terminated copy of an XPM file.  The [Scintilla documentation](https://www.scintilla.org/ScintillaDoc.html#Images) refers to these as "lines form" and "text form".  To try and tell which kind of data has been input, on a 32 bit system it's safe to look at the first 4 bytes and on a 64 bit system it's safe to look at the first 8 bytes - but I think it just picks the smaller number since it will work with both.  If those first 4 bytes are not '/' '*' ' ' 'X', then there's data sent could be in "text form" and we shouldn't look any further at the contents of that memory location.  Instead we should that memory contents as a pointer to first string of the XPM data.

I'm not 100% sure but I believe that's why the more complicated check is used in xpm.cpp instead of the simpler one you're suggesting.

vadz replied:
Do we need to support such polymorphism? Can't we only accept real `const char*` here and not either it or `const char* const*`?

NewPagodi replied:
This function is what ultimately does the processing for the SendMsg method with the SCI_REGISTERIMAGE message.  Both of these calls should work with wxSTC

```
char * somedata_xpm[] = .... (some xpm data here)
m_stc->SendMsg(SCI_REGISTERIMAGE,0,reinterpret_cast<wxIntPtr>(somedata_xpm));
```

```
const char* someXMPfile = ... (load an xpm an xmp file here)
m_stc->SendMsg(SCI_REGISTERIMAGE,0,reinterpret_cast<wxIntPtr>(someXMPfile));
```
Both of those calls will wind through several different files in both the STC and Scintilla sourced and ultimately end up in this method.

vadz replied:
OK, let's merge it like this but this is really pretty ugly and I'm almost sure that the original problem (which forced you to update this code) is somehow related to mixing up string pointers and arrays of string pointers somewhere. I'd also feel better if you could compile (and then run) the sample with ASAN and/or under Valgrind...

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -2541,20 +2521,31 @@ void ListBoxImpl::RegisterImageHelper(int type, wxBitmap& bmp)
+
     wxBitmap bmp(img);
-    RegisterImageHelper(type, bmp);
+    RegisterImageHelper(type,bmp);
```
Extremely minor, but there doesn't seem any reason to remove this space...
```suggestion
    RegisterImageHelper(type, bmp);
```

vadz commented on src/stc/ScintillaWX.cpp:
```diff
@@ -1336,6 +1338,37 @@ bool ScintillaWX::GetUseAntiAliasing() {
+            rgba[curRGBALoc++] = img.GetData()[curDataLoc++];
+            rgba[curRGBALoc++] = img.GetData()[curDataLoc++];
+            rgba[curRGBALoc++] =
+                img.HasAlpha()?img.GetAlpha()[curAlphaLoc++]:255;
```
```suggestion
                img.HasAlpha() ? img.GetAlpha()[curAlphaLoc++] : wxALPHA_OPAQUE ;
```

vadz reviewed (commented):
No real comments, if you can please address the `strncmp` remark, it would be nice because I'm reluctant to do it without any testing (even though it really shouldn't change the code meaning). Otherwise I can just apply this as is, please let me know.

vadz committed:
Update src/stc/PlatWX.cpp

Co-Authored-By: NewPagodi <NewPagodi@users.noreply.github.com>

vadz committed:
Update src/stc/ScintillaWX.cpp

Co-Authored-By: NewPagodi <NewPagodi@users.noreply.github.com>
